src/OrchardCore.Modules/OrchardCore.Queries/Sql/SqlQuerySource.cs

```diff
@@ -84,33 +103,32 @@ public async Task<IQueryResults> ExecuteQueryAsync(Query query, IDictionary<stri
                     else
                     {
                         column = rowDictionary.FirstOrDefault(kv => kv.Value is long).Key
-                            ?? rowDictionary.First().Key;
+                                 ?? rowDictionary.First().Key;
                     }
                 }
 
-                return rowDictionary.TryGetValue(column, out var documentIdObject) && documentIdObject is long documentId
-                    ? documentId
-                    : 0;
-            }).ToArray();
-
+                return (rowDictionary.TryGetValue(column, out var documentIdObject) ? documentIdObject : null) switch
+                {
+                    long longValue => longValue,
+                    ulong unsignedLongValue => (long)unsignedLongValue,
+                    int intValue => intValue, 
+                    uint unsignedIntValue => unsignedIntValue, 
+                    { } otherObject => 
+                        long.TryParse(otherObject.ToString(), CultureInfo.InvariantCulture, out var parsedValue)
+                            ? parsedValue
+                            : -1,
+                    _ => -1,
+                };
+            })
+            .Where(id => id >= 0)
+            .Distinct()
```

Comment:
I don't think we should make the result distinct here. This may alter the result someone expects from their query.